### ISSUE TITLE
Keeping of pending for delivering messages in CMD field of Delivery SBBs

### ISSUE THREAD

**Issue by vetss** [RestComm/smscgateway#98]
We keep now pending for delivery messages (when we are processing other messages for delivering to a single destination) in SmsSetCache.lstSmsSetInProcessing collection. When we need to update this collection or a message set in SmsSet for delivering then we need to make synch with usage of SmsSetCache.lstSmsSetUnderAtomicOper. This synching may be a bottleneck for productivity.

We can make the following solution:
- keep pending for delivery messages in an SBB CMP field (that contains SmsSet)
- add / remove messages to that SmsSet only inside that SBB. For adding of messages from an extrernal sources (new incoming or rerouted or obtained from database messages) we will use a new event of the SBB (and therefor all message processing will be inside SBB). This leads that will not need any synch more
- we need to implement a special code for seaching of an Activity that is responsible for delivery messages a destination and then fire an event for adding messages for the activity. Activity may be ScheduleActivity (of ScheduleRA) or NullActivity (which one ScheduleActivity/NullActivity will we use and algo of seaching of an Activity is a topic for discusion).

**Comment by faizann** [comment]
Hi

I have been looking into this as well. I see that DeliveryCommonSbb is storing SmsSet in CMP but in SchedulerRA the doInjectSmsOnFly pushes it to the processing queue. I think if we just comment out this code in doInjectSmsOnFly then that is one thing less for queue. 


```
        if (!dbOperations_C2.c2_checkProcessingSmsSet(smsSet))
            return;      
```  

I would like to get this issue done as well. If there is a design document or guideline then it could speed up the process for me.

**Comment by vetss** [comment]
@faizann 

SmsSet objects are always stored in SmsSetCache object and not stored into a CMD field. I do not have extra design document for the issue. Commenting out of that string will not fix all problems.

Before we want to fix it I need to describe one problem that prevent me of implementing the issue right now:
- if a SBB is attached to several Activities then SLEE can create two SBB objects and invoke them at the same time and if we keep SmsSet in a CMP field then say both SBB objects will take the SmsSet object from the cache then both SBB instances may make some processing at the same time that make lead of message dowble sending for example.

We have not yet decided how to overcome it (: And therefor this issue is not in progress.

**Comment by faizann** [comment]
Yeah I saw that we have to check for processingSms. Can't we use SLEE Transcations for MT just like it is done for SMPP?

**Comment by vetss** [comment]
@faizann 

"Can't we use SLEE Transcations for MT just like it is done for SMPP" - please explain more what you mean.
